### PR TITLE
Add insert, append and delete functions

### PR DESCRIPTION
**Pull request by martaiborra**
This PR addresses https://github.com/inaos/iron-array/issues/528.

### CODE REVIEW

File: caterva/caterva.c [hunk 1]
```diff
@@ -1184,7 +1184,7 @@ int caterva_meta_update(caterva_ctx_t *ctx, caterva_array_t *array,
     return CATERVA_SUCCEED;
 }
 
-int extend_shape(caterva_array_t *array, int64_t *new_shape) {
+int extend_shape(caterva_array_t *array, int64_t *new_shape, const int64_t *start) {
```
Comment (FrancescAlted) [inline caterva/caterva.c]: If we want to be consistent, `new_shape` can also be `const int64_t`.

File: caterva/caterva.c [hunk 2]
```diff
@@ -1217,84 +1217,57 @@ int extend_shape(caterva_array_t *array, int64_t *new_shape) {
...
-    }
+    array->nchunks = array->sc->nchunks;
+    free(aux);
+    free(cparams);
 
     return CATERVA_SUCCEED;
 }
 
-int shrink_shape(caterva_array_t *array, int64_t *new_shape) {
+int shrink_shape(caterva_array_t *array, int64_t *new_shape, const int64_t *start) {
```
Comment (FrancescAlted) [inline caterva/caterva.c]: Ditto

File: caterva/caterva.c [hunk 3]
```diff
@@ -1323,42 +1296,62 @@ int shrink_shape(caterva_array_t *array, int64_t *new_shape) {
...
+                   const int64_t *start) {
     CATERVA_ERROR_NULL(ctx);
     CATERVA_ERROR_NULL(array);
     CATERVA_ERROR_NULL(new_shape);
 
+    if (start != NULL) {
+        for (int i = 0; i < array->ndim; ++i) {
+            if (start[i] % array->chunkshape[i] != 0) {
+                // Chunks cannot be cut
+                CATERVA_TRACE_ERROR("start must be a multiple of chunkshape in all dims");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: For consistency, start all trace messages with an uppercase letter.  If it is the name of a variable, surround it with `` (in this case "`start` must...").

File: caterva/caterva.c [hunk 4]
```diff
@@ -1369,9 +1362,95 @@ int caterva_resize(caterva_ctx_t *ctx, caterva_array_t *array, int64_t *new_shap
...
+
+int caterva_insert(caterva_ctx_t *ctx, caterva_array_t *array, void *buffer, int64_t buffersize,
+                   const int8_t axis, int64_t insert_start) {
+
+    CATERVA_ERROR_NULL(ctx);
+    CATERVA_ERROR_NULL(array);
+    CATERVA_ERROR_NULL(buffer);
+
+    if (axis >= array->ndim) {
+        CATERVA_TRACE_ERROR("axis cannot be greater than the number of dimensions");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: Ditto.

File: caterva/caterva.c [hunk 5]
```diff
@@ -1369,9 +1362,95 @@ int caterva_resize(caterva_ctx_t *ctx, caterva_array_t *array, int64_t *new_shap
...
+    int64_t axis_size = array->itemsize;
+    int64_t buffershape[CATERVA_MAX_DIM];
+    for (int i = 0; i < array->ndim; ++i) {
+        if (i != axis) {
+            axis_size *= array->shape[i];
+            buffershape[i] = array->shape[i];
+        }
+    }
+    if (buffersize % axis_size != 0) {
+        CATERVA_TRACE_ERROR("buffersize must be multiple of the array");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: ditto

File: caterva/caterva.h [hunk 6]
```diff
@@ -651,10 +638,57 @@ int caterva_meta_update(caterva_ctx_t *ctx, caterva_array_t *array,
...
+/**
+ * Append a buffer at the end of a caterva array.
+ *
+ * @param ctx The caterva context to be used.
+ * @param array The caterva array.
+ * @param buffer The buffer where the data is stored.
+ * @param buffersize Size (in bytes) of the buffer.
+ * @param axis The axis that will be extended to append the data.
+ *
+ *
```
Comment (FrancescAlted) [inline caterva/caterva.h]: Too many empty lines.

File: tests/test_resize.c [hunk 7]
```diff
@@ -87,111 +91,128 @@ CUTEST_TEST_TEST(resize_shape) {
...
+        case 2:
+            ((int16_t *) value)[0] = (int16_t) fill_value;
+            break;
+        case 1:
+            ((int8_t *) value)[0] = fill_value;
+            break;
+        default:
+            break;
+    }
+    CATERVA_ERROR(caterva_full(data->ctx, &params, &storage, value, &src));
```
Comment (FrancescAlted) [inline tests/test_resize.c:123]: Glad that you found that one!

File: tests/test_append.c [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    /* Fill buffer with a slice from the new chunks */
+    uint8_t *res_buffer = data->ctx->cfg->alloc(buffersize);
+    CATERVA_ERROR(caterva_get_slice_buffer(data->ctx, src, start, stop, res_buffer,
+                                           shapes.buffershape, buffersize));
+
+    for (uint64_t i = 0; i < (uint64_t) buffersize / itemsize; ++i) {
+        switch (itemsize) {
+            case 8:
+                CUTEST_ASSERT("Elements are not equal!",
+                              ((uint64_t *) buffer)[i] == ((uint64_t *) res_buffer)[i]);
```
Comment (FrancescAlted) [inline tests/test_append.c:134]: Would not it be better to compare against `fill_value` too?

**Review by FrancescAlted — APPROVED** [review]
I have some (minor) suggestions.  Other than that, this looks good to me!

**Review by aleixalcacer — APPROVED** [review]
LGTM :)

File: caterva/caterva.c (hunk 1 above)
Comment (martaiborra) [reply caterva/caterva.c]: Fixed in 7bc969a51bde3550974145ac9f056b186f6d1f20

File: caterva/caterva.c (hunk 3 above)
Comment (martaiborra) [reply caterva/caterva.c]: Fixed in 7bc969a51bde3550974145ac9f056b186f6d1f20

File: caterva/caterva.c (hunk 5 above)
Comment (martaiborra) [reply caterva/caterva.c]: Fixed in 7bc969a51bde3550974145ac9f056b186f6d1f20

File: caterva/caterva.h (hunk 6 above)
Comment (martaiborra) [reply caterva/caterva.h]: Fixed in 7bc969a51bde3550974145ac9f056b186f6d1f20

**Pull request merged**